x/ibc/core/02-client/abci.go

```diff
@@ -4,11 +4,32 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/cosmos/cosmos-sdk/x/ibc/core/02-client/keeper"
 	"github.com/cosmos/cosmos-sdk/x/ibc/core/exported"
+	ibctmtypes "github.com/cosmos/cosmos-sdk/x/ibc/light-clients/07-tendermint/types"
 )
 
 // BeginBlocker updates an existing localhost client with the latest block height.
 func BeginBlocker(ctx sdk.Context, k keeper.Keeper) {
-	_, found := k.GetClientState(ctx, exported.Localhost)
+	plan, found := k.GetUpgradePlan(ctx)
+	if found {
+		// Once we are at the last block this chain will commit, set the upgraded consensus state
+		// so that IBC clients can use the last NextValidatorsHash as a trusted kernel for verifying
+		// headers on the next version of the chain.
+		// Set the time to the last block time of the current chain.
+		// In order for a client to upgrade successfully, the first block of the new chain must be committed
+		// within the trusting period of the last block time on this chain.
+		_, exists := k.GetUpgradedClient(ctx, plan.Height-1)
+		if exists && ctx.BlockHeight() == plan.Height-1 {
+			upgradedConsState := &ibctmtypes.ConsensusState{
+				Timestamp:          ctx.BlockTime(),
+				NextValidatorsHash: ctx.BlockHeader().NextValidatorsHash,
+			}
+			bz := k.MustMarshalConsensusState(upgradedConsState)
+
+			k.SetUpgradedConsensusState(ctx, plan.Height, bz)
+		}
+	}
```

Comment:
does this mean that the ibc module Begin block should run first than the x/upgrade's one?

Reply:
No don't think so. This occurs at plan.Height - 1. While the upgrade logic happens at plan.Height